### PR TITLE
[WIP][RFC]codec adapter: move generic mem alloc to component

### PR DESCRIPTION
**Pull request by juimonen**
Codec adapter has a devm_kzalloc type memory allocator to allocate
multiple chunks of memory into a list, which can be then freed without
explicit pointers in codec deletion. Move this generic mechanism to
component class.

Signed-off-by: Jaska Uimonen <jaska.uimonen@intel.com>

### CODE REVIEW

**Comment by juimonen** [comment]
@lgirdwood heavily simplified... I don't know do we need at all the step by step free of memory individually or can we just always free all memory... usually if memory allocation fails the whole component is in very bad state. I can bring back the more refined method if needed.

File: src/audio/codec_adapter/codec/cadence.c [hunk 1]
```diff
@@ -90,7 +90,7 @@ int cadence_codec_init(struct comp_dev *dev)
 
 	comp_dbg(dev, "cadence_codec_init() start");
 
-	cd = codec_allocate_memory(dev, sizeof(struct cadence_codec_data), 0);
+	cd = comp_devm_alloc(dev, sizeof(struct cadence_codec_data), 0);
 	if (!cd) {
```
Comment (dbaluta) [inline src/audio/codec_adapter/codec/cadence.c:94]: I wonder if this shouldn't be better called devm_comp_alloc(dev, ..) similar with its Linux counterpart.

**Review by lgirdwood — CHANGES_REQUESTED** [review]
Thanks @juimonen. I would make two changes here.
1) Move it to the allocator logic. i.e. all memory logic in one place.
2) Associate it with the pipeline rather than a component. We only remove componets when we delete the pipeline and we now support dynamic pipeline so make more sense at this level.

**Comment by lgirdwood** [comment]
@cujomalainey fyi

**Comment by lgirdwood** [comment]
@juimonen this should go into main branch. There are no dependencies.

**Comment by cujomalainey** [comment]
> @cujomalainey fyi

interesting, codeowners does not recurse into directories with `*` thanks for the FYI

**Comment by cujomalainey** [comment]
Thanks for moving this, its an excellent piece of code that just needs to benefit everyone 😃

**Comment by juimonen** [comment]
> Thanks @juimonen. I would make two changes here.
> 
> 1. Move it to the allocator logic. i.e. all memory logic in one place.
> 2. Associate it with the pipeline rather than a component. We only remove componets when we delete the pipeline and we now support dynamic pipeline so make more sense at this level.

1. ok
2. I'm not sure I can do that as our driver creates the pipeline component _after_ the components themselves have been created -> I don't necessarily have a pipeline where to store the memory...

File: src/audio/codec_adapter/codec/cadence.c (hunk 1 above)
Comment (juimonen) [reply src/audio/codec_adapter/codec/cadence.c:94]: yes sure. Can revisit the naming as I need to shuffle this around anyway...

**Comment by lgirdwood** [comment]
> > Thanks @juimonen. I would make two changes here.
> > 
> > 1. Move it to the allocator logic. i.e. all memory logic in one place.
> > 2. Associate it with the pipeline rather than a component. We only remove componets when we delete the pipeline and we now support dynamic pipeline so make more sense at this level.
> 
> 1. ok
> 2. I'm not sure I can do that as our driver creates the pipeline component _after_ the components themselves have been created -> I don't necessarily have a pipeline where to store the memory...

@ranj063 is doing some work here https://github.com/thesofproject/sof/pull/4113, we will need this and a kernel patch to change the pipeline new order.

So maybe part 1 is to align the allocations per module/components, and once kernel is ready we change from module to pipeline alignment, but lets get feedback from @ranj063 first wrt kernel timeline.

@RDharageswari fyi.

File: src/include/sof/audio/component.h [hunk 2]
```diff
@@ -425,6 +425,15 @@ struct comp_driver_info {
 	struct list_item list;		/**< list of component drivers */
 };
 
+/**
+ * \struct comp dev memory - used like devm_kzalloc
+ * \brief comp dev memory block - used for every memory allocated by comp
+ */
+struct comp_dev_mem {
+	void *ptr; /**< A pointr to particular memory block */
+	struct list_item mem_list; /**< list of memory allocated by codec */
```
Comment (cujomalainey) [inline src/include/sof/audio/component.h:434]: @lgirdwood would you have an issue if I built in a feature (behind a config flag) here to track amount allocated by each alloc? That way we call a function and query memory consumed by a component (and possibly identify what section each pointer is point to)

**Comment by gkbldcig** [comment]
Can one of the admins verify this patch?

File: src/include/sof/audio/component.h (hunk 2 above)
Comment (lgirdwood) [reply src/include/sof/audio/component.h:434]: @cujomalainey sound good to me.

**Comment by cujomalainey** [comment]
ping on this patch, this is something i would like to base a feature off of

**Comment by lgirdwood** [comment]
@juimonen whats' the ETA for this patch ? What's still to do ?

File: src/audio/component.c [hunk 3]
```diff
@@ -455,3 +455,57 @@ struct comp_dev *comp_make_shared(struct comp_dev *dev)
...
+{
+	struct comp_dev_mem *mem;
+	struct list_item *mem_list;
+	struct list_item *_mem_list;
+
+	/* Find which container keeps this memory */
+	list_for_item_safe(mem_list, _mem_list, &dev->mem.mem_list) {
+		mem = container_of(mem_list, struct comp_dev_mem, mem_list);
+		rfree(mem->ptr);
+		list_item_del(&mem->mem_list);
```
Comment (lyakh) [inline src/audio/component.c:506]: maybe first delete from the list, then free?

**Review by lyakh — COMMENTED** [review]
It still seems to me, that this isn't a good use of "devm." It's a clear reference to the equally named Linux kernel API, but it doesn't do the same thing. If the automated resource management is planned, we should make sure it gets implemented soon enough, not to leave this namespace confusion for indefinite time...

**Comment by juimonen** [comment]
let me work on this next week, was busy this week with other stuff.

**Comment by kv2019i** [comment]
@juimonen Over a year without activity, should this be closed?

**Pull request closed without merging**